### ISSUE TITLE
SIGBUS error with 1.2.2

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#8099]
**Describe the bug**

Customer has reported that they got an SIGBUS error with the latest patch release and running their benchmark, it looks like it is happening in the journal.

Impact: Since the customer doesn't run on kubernetes the system had to be restarted. 

> https://www.geeksforgeeks.org/segmentation-fault-sigsegv-vs-bus-error-sigbus/
> 2) Bus Error (also known as SIGBUS and is usually signal 10) occur when a process is trying to access memory that the CPU cannot physically address.In other words the memory tried to access by the program is not a valid memory address.It caused due to alignment issues with the CPU (eg. trying to read a long from an address which isn’t a multiple of 4). SIGBUS is abbrivation for “Bus Error”.
> 
> SIGBUS signal occurs in below cases,
> -> Program instructs the CPU to read or write a specific physical memory address which is not valid / Requested physical address is unrecognized by the whole computer system.
> -> Unaligned access of memory (For example, if multi-byte accesses must be 16 bit-aligned, addresses (given in bytes) at 0, 2, 4, 6, and so on would be considered aligned and therefore accessible, while addresses 1, 3, 5, and so on would be considered unaligned.)

**To Reproduce**

I think we can ask them what they exactly doing, what I know is that they have 24 partitions and 4 brokers with replication factor 4.

**Expected behavior**

No sigbus I assume?

**Log/Stacktrace**

```
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.928 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.971 [] [raft-server-2-raft-partition-partition-22] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33226084 (first log index: 33091065)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.977 [] [raft-server-2-raft-partition-partition-24] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 51869258 (first log index: 51728804)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.984 [] [raft-server-2-raft-partition-partition-10] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 34657647 (first log index: 34533361)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:33.999 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.045 [] [raft-server-2-raft-partition-partition-22] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33226084 (first log index: 33091065)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.048 [] [raft-server-2-raft-partition-partition-24] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 51869258 (first log index: 51728804)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.061 [] [raft-server-2-raft-partition-partition-10] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 34657647 (first log index: 34533361)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:34.070 [] [raft-server-2-raft-partition-partition-6] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 33400058 (first log index: 33322496)
[36mzeebe_broker2 |[0m [thread 82 also had an error]
[36mzeebe_broker2 |[0m [thread 104 also had an error][thread 91 also had an error]
[36mzeebe_broker2 |[0m 
[36mzeebe_broker2 |[0m [thread 99 also had an error]
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # A fatal error has been detected by the Java Runtime Environment:
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m #  [thread 83 also had an error]
[36mzeebe_broker2 |[0m [thread 93 also had an error]
[36mzeebe_broker2 |[0m [thread 49 also had an error][thread 48 also had an error]
[36mzeebe_broker2 |[0m 
[36mzeebe_broker2 |[0m SIGBUS (0x7) at pc=0x00007f91eb54d169, pid=8, tid=92
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # JRE version: OpenJDK Runtime Environment 18.9 (11.0.13+8) (build 11.0.13+8)
[36mzeebe_broker2 |[0m # Java VM: OpenJDK 64-Bit Server VM 18.9 (11.0.13+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
[36mzeebe_broker2 |[0m # Problematic frame:
[36mzeebe_broker2 |[0m # v  ~StubRoutines::jbyte_disjoint_arraycopy
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # Core dump will be written. Default location: /var/swiftcore/core.%e.8.%h.%t
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # An error report file with more information is saved as:
[36mzeebe_broker2 |[0m # /usr/local/zeebe/hs_err_pid8.log
[36mzeebe_broker2 |[0m Compiled method (c2) 103585123 19667   !   4       io.camunda.zeebe.journal.file.SegmentedJournalWriter::append (54 bytes)
[36mzeebe_broker2 |[0m  total in heap  [0x00007f91f4170a90,0x00007f91f4174fe8] = 17752
[36mzeebe_broker2 |[0m  relocation     [0x00007f91f4170c08,0x00007f91f4170da0] = 408
[36mzeebe_broker2 |[0m  main code      [0x00007f91f4170da0,0x00007f91f41730a0] = 8960
[36mzeebe_broker2 |[0m  stub code      [0x00007f91f41730a0,0x00007f91f4173120] = 128
[36mzeebe_broker2 |[0m  oops           [0x00007f91f4173120,0x00007f91f4173148] = 40
[36mzeebe_broker2 |[0m  metadata       [0x00007f91f4173148,0x00007f91f4173340] = 504
[36mzeebe_broker2 |[0m  scopes data    [0x00007f91f4173340,0x00007f91f4174738] = 5112
[36mzeebe_broker2 |[0m  scopes pcs     [0x00007f91f4174738,0x00007f91f4174dd8] = 1696
[36mzeebe_broker2 |[0m  dependencies   [0x00007f91f4174dd8,0x00007f91f4174e00] = 40
[36mzeebe_broker2 |[0m  handler table  [0x00007f91f4174e00,0x00007f91f4174ef0] = 240
[36mzeebe_broker2 |[0m  nul chk table  [0x00007f91f4174ef0,0x00007f91f4174fe8] = 248
[36mzeebe_broker2 |[0m Could not load hsdis-amd64.so; library not loadable; PrintAssembly is disabled
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 |[0m # If you would like to submit a bug report, please visit:
[36mzeebe_broker2 |[0m #   https://bugreport.java.com/bugreport/crash.jsp
[36mzeebe_broker2 |[0m #
[36mzeebe_broker2 exited with code 134
```

[hs_err_pid8.log](https://github.com/camunda-cloud/zeebe/files/7442605/hs_err_pid8.log)

**Environment:**
 - OS: 
 - Zeebe Version: 1.2.2
 - Configuration: 4 broker, 24 partitions, bare metal

**Comment by Zelldon** [comment]
What I can see from the customer data is that the Broker which received the SIGBUS went out of disk space and was not able to delete on several partitions data, on other partitions it was able to delete data. No exporters are configured. 


We see endless log statements of:

```
[0m 2021-10-27 14:14:17.730 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:17.730 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:17.731 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:17.731 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
```

SIGBUS message which includes CPD and jdk

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0x7) at pc=0x00007f91eb54d169, pid=8, tid=92
#
# JRE version: OpenJDK Runtime Environment 18.9 (11.0.13+8) (build 11.0.13+8)
# Java VM: OpenJDK 64-Bit Server VM 18.9 (11.0.13+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# v  ~StubRoutines::jbyte_disjoint_arraycopy
#
# Core dump will be written. Default location: /var/swiftcore/core.%e.8.%h.%t
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
---------------  S U M M A R Y ------------

Command Line: -Xmx6G -XX:+UseG1GC -XX:+ExplicitGCInvokesConcurrent -XX:MaxInlineLevel=15 -Xlog:gc*=debug:file=/usr/local/zeebe/logs/gc.log -XX:+HeapDumpOnOutOfMemoryError -XX:+ExitOnOutOfMemoryError -Xms128m --illegal-access=deny -XX:+ExitOnOutOfMemoryError -Dapp.name=broker -Dapp.pid=8 -Dapp.repo=/usr/local/zeebe/lib -Dapp.home=/usr/local/zeebe -Dbasedir=/usr/local/zeebe io.camunda.zeebe.broker.StandaloneBroker

Host: Intel(R) Xeon(R) Platinum 8270 CPU @ 2.70GHz, 12 cores, 31G, Debian GNU/Linux 11 (bullseye)
Time: Wed Oct 27 14:14:34 2021 UTC elapsed time: 103585.121013 seconds (1d 4h 46m 25s)

```

**Comment by Zelldon** [comment]
@deepthidevaki @npepinpe can we have an endless loop in deleting or why logs can be printed for same partition so often in ms ?

```
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.668 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.669 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.670 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.671 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.672 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.672 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.673 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
[36mzeebe_broker2 |[0m 2021-10-27 14:14:18.675 [] [raft-server-2-raft-partition-partition-5] DEBUG
[36mzeebe_broker2 |[0m       io.camunda.zeebe.journal.file.SegmentedJournal - No segments can be deleted with index < 49499482 (first log index: 49463401)
```

**Comment by deepthidevaki** [comment]
Deleting logs are not retried. So I expect this log only once when a new snapshot is taken or received.

**Comment by deepthidevaki** [comment]
Oh.. wait!  May be we try to compact when raft detects it is OOD? May be in LeaderAppender or FollowerRole.

**Comment by deepthidevaki** [comment]
:facepalm: 

https://github.com/camunda-cloud/zeebe/blob/7e33b488a1fed1c8fbf5c2fffdede1275007795f/atomix/cluster/src/main/java/io/atomix/raft/roles/PassiveRole.java#L666-L669

**Comment by deepthidevaki** [comment]
@Zelldon FYI OOD may be due to this issue https://github.com/camunda-cloud/zeebe/issues/7767

**Comment by npepinpe** [comment]
Please timebox some effort to root cause it, then put it back in triage when we have more information on this.

**Comment by Zelldon** [comment]
SIGBUS seems of happen with java related to full `/tmp` folders

 * https://stackoverflow.com/questions/18136108/java-jvm-crashes-before-running-my-program
 * https://confluence.atlassian.com/confkb/java-vm-dies-with-sigbus-0x7-when-temp-directory-is-full-on-linux-815584538.html

**Comment by npepinpe** [comment]
From the looks, it's usually because it's memory mapping the files in /tmp - which succeeds, but when it tries to write it fails. I'm not sure why it's SIGBUS and not the `InternalError` we've previously observed - it could be because in this case, the JVM is not using its own `MappedByteBuffer` API and using something else? I would be curious to try a small test which files a disk, maps a file and then try to write and see if I get a SIGBUS, or if this only happens for JVM reports written to /tmp :thinking:

**Comment by Zelldon** [comment]
Could it be that if the disk is full and we end in the endless loop like https://github.com/camunda-cloud/zeebe/issues/8103 we print so many logs that we write tmp full and end in a SIGBUS because the JVM fails with the full `/tmp`?

**Comment by npepinpe** [comment]
Could be! I would feel a bit better about this than knowing we have to handle SIGBUS with our own memory mapped files :smile:

**Comment by Zelldon** [comment]
Was able to reproduce it.


```
{"severity":"DEBUG","logging.googleapis.com/sourceLocation":{"function":"storeExporterPositions","file":"ExporterPositionsDistributionService.java","line":44},"message":"Received new exporter state {MetricsExporter=674628}","serviceContext":{"service":"zeebe-broker","version":"zell-disk-test"},"context":{"threadId":27,"partitionId":"3","threadPriority":5,"loggerName":"io.camunda.zeebe.broker.exporter","threadName":"Broker-1-zb-fs-workers-2","actor-name":"Broker-1-Exporter-3"},"timestampSeconds":1635775939,"timestampNanos":140792000}
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0x7) at pc=0x00007fa2e754d162, pid=7, tid=62
#
# JRE version: OpenJDK Runtime Environment 18.9 (11.0.13+8) (build 11.0.13+8)
# Java VM: OpenJDK 64-Bit Server VM 18.9 (11.0.13+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# v  ~StubRoutines::jbyte_disjoint_arraycopy
#
# Core dump will be written. Default location: /core.%e.7.%t
#
# An error report file with more information is saved as:
# /usr/local/zeebe/data/zeebe_error7.log
[thread 60 also had an error]
[thread 61 also had an error]
Compiled method (c2)  660342 9894       4       io.camunda.zeebe.journal.file.record.SBESerializer::writeData (144 bytes)
 total in heap  [0x00007fa2ef710d90,0x00007fa2ef712d20] = 8080
 relocation     [0x00007fa2ef710f08,0x00007fa2ef710fd8] = 208
 main code      [0x00007fa2ef710fe0,0x00007fa2ef711fa0] = 4032
 stub code      [0x00007fa2ef711fa0,0x00007fa2ef711fc8] = 40
 oops           [0x00007fa2ef711fc8,0x00007fa2ef711fe0] = 24
 metadata       [0x00007fa2ef711fe0,0x00007fa2ef7120e0] = 256
 scopes data    [0x00007fa2ef7120e0,0x00007fa2ef712848] = 1896
 scopes pcs     [0x00007fa2ef712848,0x00007fa2ef712c48] = 1024
 dependencies   [0x00007fa2ef712c48,0x00007fa2ef712c58] = 16
 handler table  [0x00007fa2ef712c58,0x00007fa2ef712c88] = 48
 nul chk table  [0x00007fa2ef712c88,0x00007fa2ef712d20] = 152
Could not load hsdis-amd64.so; library not loadable; PrintAssembly is disabled
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
# 
```

[zeebe-values.yaml.txt](https://github.com/camunda-cloud/zeebe/files/7453827/zeebe-values.yaml.txt)


After failing once it seems to fail in a loop, can't access anymore the pod
```
  State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    134
      Started:      Mon, 01 Nov 2021 15:15:02 +0100
      Finished:     Mon, 01 Nov 2021 15:15:30 +0100
    Ready:          False
```

**Comment by npepinpe** [comment]
Are you filling the data volume or /tmp?

**Comment by Zelldon** [comment]
Please see above, the `usr/local/zeebe/data`

**Comment by npepinpe** [comment]
So writing to a memory mapped file when there is no space doesn't produce an `InternalError` but instead a SIGBUS? How come last time we observed it was an internal error?

Any rate, the fix is then to pre-allocate files. We've already discussed this, pre-allocate the files before memory mapping them to ensure we have enough disk space and don't need to deal with weird errors. See https://github.com/camunda-cloud/zeebe/issues/7607.
